### PR TITLE
fix: add update manager config in updateFunc.

### PR DESCRIPTION
**Pull request by kycheng**
# Changes

1. add update manager config in updateFunc.
2. remove version generate.

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [`spec` PR link](https://github.com/katanomi/spec) included
- [x] Follows the [commit message standard](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md#commits)
- [ ] Meets the [contributing guidelines](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md) (including
  functionality, content, code)
- [x] Test cases with documentation and functionality works as expected using current and related github repos (MUST deploy and check)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

### CODE REVIEW

File: apis/meta/v1alpha1/buildmetadata_types.go [hunk 1]
```diff
@@ -76,11 +76,11 @@ type BuildRunGitStatus struct {
 
 	// VersionPhase is the phase on the versionscheme that matches this git revision.
 	// +optional
-	VersionPhase string `json:"versionPhase,omitempty"`
+	VersionPhase string `json:"versionPhase,omitempty" variable:"-"`
```
Comment (kycheng) [inline apis/meta/v1alpha1/buildmetadata_types.go:79]: 版本策略的变量统一走build.status 生成。

File: config/manager.go [hunk 2]
```diff
@@ -98,15 +98,30 @@ func (manager *Manager) GetConfig() *Config {
...
 	defaultValue := defaultFeatureValue[flag]
-	if manager == nil {
+	if manager == nil || manager.Config == nil {
 		return defaultValue
 	}
 
-	if value, ok := manager.Data[flag]; ok {
+	manager.lock.Lock()
+	value, ok := manager.Data[flag]
+	manager.lock.Unlock()
```
Comment (l-qing) [inline config/manager.go:123]: 一般是用 defer ？

File: config/feature_watch.go [hunk 3]
```diff
@@ -99,6 +99,9 @@ func predicatesUpdateFunc(manager *Manager, featureChanged HasFeatureChangedFunc
 			featureChanged = defaultFeatureChanged
 		}
 
+		// Update configuration files manually. the update of the watch may be later than the time trigger.
+		manager.UpdateConfig(new)
+
```
Comment (danielfbm) [inline config/feature_watch.go:104]: predicate里面加这个功能不合适

**Review by danielfbm — CHANGES_REQUESTED** [review]
留了评论

File: config/feature_watch.go [hunk 4]
```diff
@@ -67,6 +67,11 @@ func enqueueRequestsConfigMapFunc(ctx context.Context, listFunc ListByFeatureFla
 			return
 		}
 
+		// Update configuration files manually. the update of the watch may be later than the time trigger.
+		if manager != nil && manager.Informer != nil {
+			manager.Informer.OnChange(configMap)
+		}
```
Comment (danielfbm) [inline config/feature_watch.go:73]: 不应该更新manager的数据，它自己有更新机制，这里保持只读操作

File: config/feature_watch.go [hunk 5]
```diff
@@ -51,10 +52,34 @@ func (manager *Manager) WatchFeatureFlagChanged(ctx context.Context, listFunc Li
...
+	oldVersion := manager.GetResourceVersion()
+	timeout, _ := manager.GetFeatureFlag(ConfigUpdateTimeoutKey).AsInt()
+
+	timeoutDuration := time.After(time.Duration(time.Millisecond * time.Duration(timeout)))
+	ticker := time.NewTicker(time.Microsecond * 10)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if cm.ResourceVersion == manager.GetResourceVersion() {
```
Comment (danielfbm) [inline config/feature_watch.go:72]: cm.ResourceVersion >= manager.GetResourceVersion ? 这里得把它当int来处理

File: config/feature_watch.go [hunk 6]
```diff
@@ -95,6 +120,10 @@ func predicatesUpdateFunc(manager *Manager, featureChanged HasFeatureChangedFunc
 			return false
 		}
 
+		// wait for the manager to update to the latest configuration.
+		// Ignore the rare case of timeout and try to trigger reconcile.
+		manager.WaitWatchResourceApply(ctx, new)
```
Comment (danielfbm) [inline config/feature_watch.go:125]: 这种操作放在 event handler 那种函数合适，这个predicate可以理解为一个过滤条件而已，那个函数可以处理timeout的情况，比如timeout就不reconcile或者reconcile所有的？

**Review by danielfbm — APPROVED** [review]

**Pull request merged**